deep_quoridor/src/quoridor.py

```diff
@@ -0,0 +1,188 @@
+        if orientation == WallOrientation.VERTICAL:
+            return self._grid[position[0] * 2 + 1 : position[0] * 2 + 4, position[1] * 2 + 1]
+        else:
+            return self._grid[position[0] * 2 + 1, position[1] * 2 + 1 : position[1] * 2 + 4]
+
+    def get_walls_between(self, position1: tuple[int, int], position2: tuple[int, int]) -> bool:
+        """
+        Return a sequence of occupancy values (both wall and player positions) between two player positions
+
+        The first element of the sequence will be player, then wall, etc.
+        The last element of the sequence will be the player position at position2.
+        """
+        start = (position1[0] * 2, position1[1] * 2)
+        destination = (position2[0] * 2, position2[1] * 2)
+        return self._grid[start[0] : destination[0] + 1, start[1] : destination[1] + 1]
+
+    def __str__(self):
+        return "\n".join("".join("X" if cell else "." for cell in row) for row in self._grid)
+
+
+class Quoridor:
+    def __init__(self, board_size: int, max_walls: int):
+        self.board_size = board_size
+        self.max_walls = max_walls
+        self.walls = np.zeros((board_size - 1, board_size - 1, 2), dtype=np.int8)
+        self.player_positions = [(0, board_size // 2), (board_size - 1, board_size // 2)]
+        self.walls_remaining = [self.max_walls, self.max_walls]
+        self.current_player = 0
+
+        # Occupancy grid which has both walls and possible pawn positions.
+        # Occupied pawn and wall positions are marked as True.
+        self._combined_occ_grid = CombinedGrid(board_size)
+
+    def step(self, action: Action, validate: bool = True):
+        """
+        Execute the given action and update the game state.
+
+        Turning off validation can save some computation if you're sure the action is valid,
+        and also makes it easier to teleport players around the board for testing purposes.
+        """
+        if validate and not self.is_action_valid(action):
+            raise ValueError("Invalid action")
+
+        if isinstance(action, MoveAction):
+            # Move the pawn, mark previous position as empty, and mark new position as occupied.
+            self._combined_occ_grid.set_player_cell(self.player_positions[self.current_player], False)
+            self.player_positions[self.current_player] = action.destination
+            self._combined_occ_grid.set_player_cell(self.player_positions[self.current_player], True)
+        elif isinstance(action, WallAction):
+            self._combined_occ_grid.set_wall_cells(action.position, action.orientation, True)
+            # TODO: Check that the wall doesn't block any player from reaching their goal.
+            self.walls_remaining[self.current_player] -= 1
+        else:
+            raise ValueError("Invalid action type")
+
+        self.current_player = 1 - self.current_player
+
+    def is_action_valid(self, action: Action):
```

Comment:
This is duplicating the checks in `_get_action_mask`, since they are complex they should be coded only once.
`_get_action_mask` could call on this for each possibility, but the pawn movement is more straightforward to code the other way around, like it's now in `_get_action_mask`.